build/esm/TokenRefresher.js

```diff
@@ -0,0 +1,143 @@
+                catch (error) {
+                    this.adapter.log.error(`Cannot parse tokens: ${error}`);
+                    this.accessToken = undefined;
+                }
+            }
+        }
+    }
+    /** Returns the access token if it is valid and not expired.*/
+    async getAccessToken() {
+        await this.readyPromise;
+        if (!this.accessToken?.access_token) {
+            this.adapter.log.error(`No tokens for ${this.name} found`);
+            return undefined;
+        }
+        if (!this.accessToken.access_token_expires_on ||
+            new Date(this.accessToken.access_token_expires_on).getTime() < Date.now()) {
+            this.adapter.log.error('Access token is expired. Please make a authorization again');
+            return undefined;
+        }
+        return this.accessToken.access_token;
+    }
+    async refreshTokens() {
+        if (this.refreshTokenTimeout) {
+            this.adapter.clearTimeout(this.refreshTokenTimeout);
+            this.refreshTokenTimeout = undefined;
+        }
+        if (!this.accessToken?.refresh_token) {
+            this.adapter.log.error(`No tokens for ${this.name} found`);
+            return;
+        }
+        if (!this.accessToken.access_token_expires_on ||
+            new Date(this.accessToken.access_token_expires_on).getTime() < Date.now()) {
+            this.adapter.log.debug('Access token is expired. Retrying to refresh tokens...');
+        }
+        let expiresIn = new Date(this.accessToken.access_token_expires_on).getTime() - Date.now() - 180_000;
+        // If expiration is in less than 3 minutes, refresh the token
+        if (expiresIn <= 0) {
+            // Refresh token
+            try {
+                const response = await axios.post(this.url, this.accessToken);
+                if (response.status !== 200) {
+                    this.adapter.log.error(`Cannot refresh tokens: ${response.statusText}`);
+                    return;
+                }
+                this.accessToken = response.data;
+            }
+            catch (error) {
+                this.adapter.log.error(`Cannot refresh tokens: ${error}`);
+            }
+            if (this.accessToken) {
+                this.accessToken.access_token_expires_on = new Date(Date.now() + this.accessToken.expires_in * 1_000).toISOString();
+                expiresIn = new Date(this.accessToken.access_token_expires_on).getTime() - Date.now() - 180_000;
+                await this.adapter.setState(this.stateName, JSON.stringify(this.accessToken), true);
+                this.adapter.log.debug(`Tokens for ${this.name} updated`);
+            }
+            else {
+                // Try again in 10 minutes
+                expiresIn = 600_000; // 10 minutes
+                this.adapter.log.error(`No tokens for ${this.name} could be refreshed`);
+            }
+        }
+        // no longer than 10 minutes, as longer timer could be not reliable
+        if (expiresIn > 600_000) {
```

Comment:
You really want one call to your service every 10 minutes for each user that uses that? usually tokens have a lifetime of 1+ hours up to days and such ... maybe it is ok to have a shorter "check" timer, but not for refresh calls.

Reply:
It is only check if accessToken expired without asking the cloud